cmd/thanos/receive.go

```diff
@@ -97,13 +134,45 @@ func runReceive(
 		MaxBlockDuration:  model.Duration(time.Hour * 2),
 	}
 
+	var hashring receive.Hashring
+	// Run File Service Discovery and check for updates.
+	if fileSD != nil {
+		ctxRun, cancelRun := context.WithCancel(context.Background())
+		fileSDUpdates := make(chan []*targetgroup.Group)
+		go fileSD.Run(ctxRun, fileSDUpdates)
```

Comment:
Not sure if this is a common thing in the discovery package, but what about checking errors here? Seems like this, if at all, would be an upstream improvement.

Reply:
What errors are you referring to? Run returns no errors https://github.com/prometheus/prometheus/blob/master/discovery/file/file.go#L221

Reply:
I thought about the upstream function not returning some errors, but this seems good.

Reply:
should this be a separate run group though?

Reply:
Hmm putting it in a separate run group means it is async and we can’t wait to receive a target group on the chan until after `g.Run()`. This means we can’t construct a hashring directly with a target group but instead the hashring needs to accept a `Discoverer`. That’s doable, just means some rearchitecting.